mistralrs-core/src/utils/gguf_metadata.rs

```diff
@@ -0,0 +1,125 @@
+use akin::akin;
+use anyhow::ensure;
+use anyhow::Result;
+use candle_core::quantized::gguf_file;
+use std::collections::HashMap;
+use tracing::warn;
+
+pub struct ContentMetadata<'a> {
+    pub path_prefix: String,
+    pub metadata: &'a HashMap<String, gguf_file::Value>,
+}
+
+impl ContentMetadata<'_> {
+    // Retrieve a prop the struct needs by querying the metadata content:
+    pub fn get_value<T: TryFromValue>(&self, field_name: &str) -> Result<T, anyhow::Error> {
+        let prop_key = format!("{prefix}.{field_name}", prefix = self.path_prefix);
+        let value = self.metadata.get(&prop_key).cloned();
+
+        // Unwrap the inner value of the `Value` enum via trait method,
+        // otherwise format error with prop key as context:
+        value
+            .try_value_into()
+            .or_else(|e| anyhow::bail!("`{prop_key}` `{e}`"))
+    }
+
+    // Fail early - Catch all missing mandatory keys upfront:
+    pub fn has_required_keys(&self, fields: &[&str]) -> Result<()> {
+        let mut all_props_are_present = true;
+
+        for field_name in fields {
+            let prop_key = format!("{prefix}.{field_name}", prefix = self.path_prefix);
+
+            if !self.metadata.contains_key(&prop_key) {
+                all_props_are_present = false;
+                warn!("Expected GGUF metadata to have key: `{prop_key}`");
+            }
+        }
+
+        ensure!(all_props_are_present, "Tokenizer is missing required props");
+        Ok(())
+    }
+
+    // Reference: https://github.com/ggerganov/ggml/blob/master/docs/gguf.md#required
+    pub fn verify_arch(&self, expected_arch: &str) -> Result<()> {
+        let actual_arch: String = self
+            .metadata
+            .get("general.architecture")
+            .cloned()
+            .try_value_into()?;
+
+        anyhow::ensure!(
+            actual_arch == expected_arch,
+            "Expected `{expected_arch}` architecture, got `{actual_arch}`."
+        );
```

Comment:
Your `verify_sanity_gguf()` method was migrated to this one.

Reply:
Sounds great.